apisix/discovery/nacos.lua

```diff
@@ -0,0 +1,306 @@
+        weight = {type = 'integer', minimum = 1, default = 100},
+        timeout = {
+            type = 'object',
+            properties = {
+                connect = {type = 'integer', minimum = 1, default = 2000},
+                send = {type = 'integer', minimum = 1, default = 2000},
+                read = {type = 'integer', minimum = 1, default = 5000},
+            },
+            default = {
+                connect = 2000,
+                send = 2000,
+                read = 5000,
+            }
+        },
+    },
+    required = {'host'}
+}
+
+
+local _M = {}
+
+
+local function request(request_uri, path, body, method, basic_auth)
+    local url = request_uri .. path
+    log.info('request url:', url)
+    local headers = {}
+    headers['Accept'] = 'application/json'
+
+    if basic_auth then
+        headers['Authorization'] = basic_auth
+    end
+
+    if body and 'table' == type(body) then
+        local err
+        body, err = core.json.encode(body)
+        if not body then
+            return nil, 'invalid body : ' .. err
+        end
+        headers['Content-Type'] = 'application/json'
+    end
+
+    local httpc = http.new()
+    local timeout = local_conf.discovery.nacos.timeout
+    local connect_timeout = timeout.connect
+    local send_timeout = timeout.send
+    local read_timeout = timeout.read
+    log.info('connect_timeout:', connect_timeout, ', send_timeout:', send_timeout,
+             ', read_timeout:', read_timeout)
+    httpc:set_timeouts(connect_timeout, send_timeout, read_timeout)
+    local res, err = httpc:request_uri(url, {
+        method = method,
+        headers = headers,
+        body = body,
+        ssl_verify = true,
+    })
+    if not res then
+        return nil, err
+    end
+
+    if not res.body or res.status ~= 200 then
+        return nil, 'status = ' .. res.status
+    end
+
+    local json_str = res.body
+    local data, err = core.json.decode(json_str)
+    if not data then
+        return nil, err
+    end
+    return data
+end
+
+
+local function get_url(request_uri, path)
+    return request(request_uri, path, nil, 'GET', nil)
+end
+
+
+local function post_url(request_uri, path, body)
+    return request(request_uri, path, body, 'POST', nil)
+end
+
+
+local function get_token_param(base_uri, username, password)
+    if not username or not password then
+        return ''
+    end
+
+    local args = { username = username, password = password}
+    local data, err = post_url(base_uri, auth_path .. '?' .. ngx.encode_args(args), nil)
+    if err then
+        log.error('nacos login fail:', username, ' ', password, ' desc:', err)
+        return nil, err
+    end
+    return '&accessToken=' .. data.accessToken
+end
+
+
+local function get_base_uri()
+    local host = local_conf.discovery.nacos.host
+    -- TODO Add health check to get healthy nodes.
+    local url = host[math_random(#host)]
+    local auth_idx = str_find(url, '@', 1, true)
```

Comment:
The `core.string.find` is already the plain string find... 😅 
Not all the suggestion from the reviewer is correct.